### PR TITLE
Improve server startup sequencing.

### LINKED ISSUE #1369: Same game settings, different player starting locations depending on platform
### Environment

* **FreeOrion Version**: 4644c927ccf8a02f36e59326832d7cfc31e31595
* **Operating System**: Windows, OSX, Linux
* **Fetched as**
  * Compiled from source

### Description

When starting a default quickstart game, assignment of home systems to empires differs depending on platform. I've tested this on all three platforms: Windows (XP), OSX (10.10) and Linux (Kubuntu 16.04). For some strange reason I can't insert them here, so I opened a thread on the forum where I posted the screenshots:

http://www.freeorion.org/forum/viewtopic.php?f=28&t=10434

### Expected Result

The map as generated by the universe generation scripts should be identical across all platforms when using identical game settings.

### Steps to reproduce

Just start a default quickstart game on different platforms and compare the results.

### PR DESCRIPTION
**Pull request by LGM-Doyle**
This may fix #1369.  However, I don't have the multiple OSs to test it for myself.

This PR changes the startup sequencing for new games as explained below.

The previous startup sequencing for a new game was: start the python engine, start accepting network connections, generate the universe, spawn any AIs, wait for the AIs to callback, and finally send the game start data to all of the players.  

This PR changes the sequence to: start accepting network connections, spawn the AIs, then start the python engine and generate the universe in parallel, wait for the AIs to callback and universe generation to complete,  and finally send the game start data to the players.

The PR provides the following benefits:
- the network connection starts responding to requests up a few seconds earlier.
- python startup and universe generation can overlap AI startup resulting in faster startup.
- the AI ids are assigned independent of their callback order and their shared_ptr sort order.  This is the change that might fix #1369.

### CODE REVIEW

**Comment by geoffthemedio** [comment]
builds, seems to run OK on windows

**Comment by Vezzra** [comment]
> the AI ids are assigned independent of their callback order and their shared_ptr sort order. This is the change that might fix #1369.

Unfortunately not. The problem isn't the order of player/empires in the player setup data dict passed into `create_universe`, as I found out in the meantime, but the order of the list of home systems compiled by the home system selection algorithm (I've already posted a more detailed explanation of what I found [here](https://github.com/freeorion/freeorion/issues/1369#issuecomment-285738375), you might want to take a look at this btw).

Anyway, even if this doesn't fix #1369, it's still a very good improvement IMO.

File: server/ServerApp.cpp [hunk 1]
```diff
@@ -231,6 +226,33 @@ void ServerApp::CreateAIClients(const std::vector<PlayerSetupData>& player_setup
     // binary / executable to run for AI clients
     const std::string AI_CLIENT_EXE = AIClientExe();
 
+
+    // TODO: add other command line args to AI client invocation as needed
+    std::vector<std::string> args, arg;
+    args.push_back("\"" + AI_CLIENT_EXE + "\"");
+    args.push_back("place_holder");
+    size_t player_pos = args.size()-1;
+    std::stringstream maxAggrStr;
```
Comment (geoffthemedio) [inline server/ServerApp.cpp]: not original to this PR, but should be renamed `max_aggr_str`

File: server/ServerApp.cpp [hunk 2]
```diff
@@ -606,66 +586,35 @@ void ServerApp::NewMPGameInit(const MultiplayerLobbyData& multiplayer_lobby_data
...
+
+    NewGameInitConcurrentWithJoiners(multiplayer_lobby_data, psds);
+    NewGameInitVerifyJoiners(multiplayer_lobby_data, psds);
 }
 
-void ServerApp::NewGameInit(const GalaxySetupData& galaxy_setup_data,
-                            const std::map<int, PlayerSetupData>& player_id_setup_data) {
-    DebugLogger() << "ServerApp::NewGameInit";
+void ServerApp::NewGameInitConcurrentWithJoiners(const GalaxySetupData& galaxy_setup_data,
+                            const std::vector<PlayerSetupData>& player_setup_data) {
```
Comment (geoffthemedio) [inline server/ServerApp.cpp]: weird indenting... move { to newline probly

File: server/ServerApp.cpp [hunk 3]
```diff
@@ -750,8 +692,100 @@ void ServerApp::NewGameInit(const GalaxySetupData& galaxy_setup_data,
 
     m_universe.UpdateStatRecords();
 
+}
+
+void ServerApp::NewGameInitVerifyJoiners(const GalaxySetupData& galaxy_setup_data,
+                                         const std::vector<PlayerSetupData>& player_setup_data) {
```
Comment (geoffthemedio) [inline server/ServerApp.cpp]: { to newline

File: server/ServerApp.cpp [hunk 4]
```diff
@@ -750,8 +692,100 @@ void ServerApp::NewGameInit(const GalaxySetupData& galaxy_setup_data,
...
+    m_galaxy_setup_data = galaxy_setup_data;
+
+    // associate player IDs with player setup data.  the player connection with
+    // id == m_networking.HostPlayerID() should be the human player in
+    // PlayerSetupData.  AI player connections are assigned one of the remaining
+    // PlayerSetupData entries that is for an AI player.
+
+    std::map<int, PlayerSetupData> player_id_setup_data;
+
+    bool host_is_human(false);
```
Comment (geoffthemedio) [inline server/ServerApp.cpp]: prefer {false} ?

File: server/ServerApp.cpp [hunk 5]
```diff
@@ -750,8 +692,100 @@ void ServerApp::NewGameInit(const GalaxySetupData& galaxy_setup_data,
...
+        return;
+    }
+
+    if (m_networking.HostPlayerID() >= player_id_setup_data.size()) {
+        ErrorLogger() << "NewGameInitVerifyJoiners : Host id " << m_networking.HostPlayerID()
+                      << " is larger than the number of players = " << player_id_setup_data.size() << ".";
+        return;
+    }
+
+    if (player_id_setup_data[m_networking.HostPlayerID()].m_client_type != Networking::CLIENT_TYPE_HUMAN_PLAYER)
```
Comment (geoffthemedio) [inline server/ServerApp.cpp]: this line is a bit long, and the { is on a newline anyway, so could split

File: server/ServerApp.h [hunk 6]
```diff
@@ -186,10 +186,14 @@ class ServerApp : public IApp {
...
 
-    /** Intializes single player game universe, sends out initial game state to
-      * clients, and signals clients to start first turn */
+    /** Intializes single player game universe.*/
     void    NewSPGameInit(const SinglePlayerSetupData& single_player_setup_data);
 
+    /** Verifies single player game AIs are compatible with already
+      * created universe. Sends out initial game state to
+      * clients, and signals clients to start first turn */
+    void    VerifySPGameAIs(const SinglePlayerSetupData& single_player_setup_data);
```
Comment (geoffthemedio) [inline server/ServerApp.h]: comment explains it does more than the name suggests

File: server/ServerFSM.cpp [hunk 7]
```diff
@@ -1047,27 +1047,32 @@ WaitingForSPGameJoiners::WaitingForSPGameJoiners(my_context c) :
...
     if (m_single_player_setup_data->m_new_game) {
-        // DO NOTHING
-
         // for new games, single player setup data contains full m_players
         // vector, so can just use the contents of that to create AI
         // clients
+        for (auto& psd : players) {
+            if (psd.m_player_name.empty()) {
+                if (psd.m_client_type == Networking::CLIENT_TYPE_AI_PLAYER)
+                    psd.m_player_name = "AI_" + boost::lexical_cast<std::string>(player_num++);
```
Comment (geoffthemedio) [inline server/ServerFSM.cpp]: prefer `std::to_string` for built-in types

File: server/ServerFSM.h [hunk 8]
```diff
@@ -195,10 +195,10 @@ struct WaitingForSPGameJoiners : sc::state<WaitingForSPGameJoiners, ServerFSM> {
     sc::result react(const LoadSaveFileFailed& u);
     sc::result react(const Error& msg);
 
-    std::shared_ptr<SinglePlayerSetupData> m_single_player_setup_data;
+    std::shared_ptr<SinglePlayerSetupData>   m_single_player_setup_data;
     std::vector<PlayerSaveGameData>          m_player_save_game_data;
-    std::shared_ptr<ServerSaveGameData> m_server_save_game_data;
-    std::set<std::string>                    m_expected_ai_player_names;
+    std::shared_ptr<ServerSaveGameData>      m_server_save_game_data;
+    std::map<std::string, int>               m_expected_ai_player_ids;
```
Comment (geoffthemedio) [inline server/ServerFSM.h]: maybe call it `m_expected_ai_names_and_player_ids` or such?

File: server/ServerApp.cpp (hunk 4 above)
Comment (LGM-Doyle) [reply server/ServerApp.cpp]: fixed.

File: server/ServerApp.cpp (hunk 5 above)
Comment (LGM-Doyle) [reply server/ServerApp.cpp]: fixed.

File: server/ServerApp.h (hunk 6 above)
Comment (LGM-Doyle) [reply server/ServerApp.h]: I broke it into two functions, `bool VerifySPGameAIs()` and `SendNewGameStartMessages()` whose functions are more clearly implied by their names.

File: server/ServerFSM.cpp (hunk 7 above)
Comment (LGM-Doyle) [reply server/ServerFSM.cpp]: fixed.

File: server/ServerFSM.h (hunk 8 above)
Comment (LGM-Doyle) [reply server/ServerFSM.h]: fixed.

**Pull request merged**